### PR TITLE
[react-form] include `dynamicLists` values in `onSubmit` type

### PR DESCRIPTION
**Pull request by sylvhama**
## Description

Follow up of https://github.com/Shopify/quilt/pull/1828

We noticed that `onSubmit` does not include the values from the dynamic lists.

Thanks to that PR, when using `dynamicLists` you can do something like this:
```ts
useForm<typeof fields, typeof dynamicLists>({...})
```
to make sure `onSubmit` will infer the values types correctly.

I have this [codesandbox](https://codesandbox.io/s/affectionate-snow-69xsx?file=/src/App.tsx) that contains a hardcoded version of `react-form`. Nothing is rendered, the goal is to hover your mouse on variables and see their types.

Screen Shot 2021-04-23 at 5 26 10 PM

Before the fix TS would not include dynamic lists values in `fieldValues`.

## Type of change

- [x] react-form Patch: Bug (non-breaking change which fixes an issue)

## Checklist

- [x] I have added a changelog entry, prefixed by the type of change noted above (Documentation fix and Test update does not need a changelog as we do not publish new version)

### CODE REVIEW

File: packages/react-form/src/types.ts [hunk 1]
```diff
@@ -124,10 +133,8 @@ export interface FieldBag {
   [key: string]: FieldOutput<any>;
 }
 
-export type DynamicListOutput<V extends object> = DynamicList<V>;
```
Comment (sylvhama) [inline packages/react-form/src/types.ts:127]: I think this is useless 🤔

File: packages/react-form/src/types.ts [hunk 2]
```diff
@@ -74,13 +74,22 @@ export interface FormWithoutDynamicListsInput<T extends FieldBag> {
...
+  fields: T;
   dynamicLists: D;
+  onSubmit?: SubmitHandler<
+    FormMapping<T, 'value'> & FormMapping<DynamicListMapping<D>, 'value'>
+  >;
+  makeCleanAfterSubmit?: boolean;
 }
 
-export interface FormInput<T extends FieldBag, D extends DynamicListBag>
-  extends FormWithoutDynamicListsInput<T> {
```
Comment (sylvhama) [inline packages/react-form/src/types.ts:82]: I removed `extends` cause it makes it hard to follow.

**Review by oluwatimio — APPROVED** [review]
This looks really good!

File: packages/react-form/README.md [hunk 3]
```diff
@@ -698,88 +731,65 @@ You can choose to initialize the list with an existing number of cards or no car
         value={field.cvv.value}
         onChange={field.cvv.onChange}
       />
-      <div>
-        <Button onClick={() => removeItem(index)}>Remove</Button>
-      </div>
```
Comment (sylvhama) [inline packages/react-form/README.md:703]: I removed the `div` to reduce the density of the component tree.

File: packages/react-form/README.md [hunk 4]
```diff
@@ -698,88 +731,65 @@ You can choose to initialize the list with an existing number of cards or no car
...
   },
   onSubmit: async fieldValues => {
     console.log(fieldValues);
-    return {status: 'success'};
+    return submitSuccess();
   },
 });
 ```
 
-A code sandbox of `useDynamicList` in action can be seen [here](https://codesandbox.io/s/hungry-rubin-exrkz?fontsize=14&hidenavigation=1&theme=dark)
```
Comment (oluwatimio) [inline packages/react-form/README.md:779]: I am not sure we should remove the sandbox as it helps people visualize how the dynamic list works

Comment (sylvhama) [reply packages/react-form/README.md:779]: If you think something i missing in the README then we can add it tho

File: packages/react-form/src/hooks/form.ts [hunk 5]
```diff
@@ -27,18 +27,14 @@ import {useDynamicListDirty, useDynamicListReset} from './list';
...
  *
  *  function MyComponent() {
- *   const { fields, submit, submitting, dirty, reset, submitErrors, dynamicLists} = useForm({
+ *   const { fields, submit, submitting, dirty, reset, submitErrors } = useForm({
  *     fields: {
  *       title: useField('some default title'),
  *     },
- *     dynamicLists: {
- *       customerPaymentMethods: useDynamicList<Card>([{cardNumber: '4242 4242 4242 4242', cvv: '422'}], emptyCardFactory)
- *     }
```
[the remark this answers is not in the document]
Comment (oluwatimio) [reply packages/react-form/src/hooks/form.ts:36]: 🤔 , shouldnt we somehow show the dynamicList param in action here? `useList` might not be there because it goes into the fields param

File: packages/react-form/README.md [hunk 6]
```diff
@@ -638,54 +638,87 @@ function MyComponent() {
...
+
 const emptyCardFactory = (): Card => ({
+  id: new Date.now().toString(),
   cardNumber: '',
   cvv: '',
 });
 
 const {fields, addItem, removeItem, moveItem, reset, dirty} = useDynamicList(
-  [{cardNumber: '4242 4242 4242 4242', cvv: '000'}],
+  [],
```
[the remark this answers is not in the document]
Comment (sylvhama) [reply packages/react-form/README.md:663]: True, that's why on L701 we have a dedicated example for that
> You can choose to initialize the list with no card (same as above) or with existing cards coming from your database for example.

File: packages/react-form/README.md [hunk 7]
```diff
@@ -638,54 +638,87 @@ function MyComponent() {
 
 #### `useDynamicList()`
 
-This offers the same functionality as useList. `useDynamicList` is a hook that adds on `useList` by adding the ability to dynamically add and remove list items. The same way you would utilize `useList` is the same way you would utilize `useDynamicList` except some differences such as the `addItem`, and the `removeItem` function. We also have to pass in a factory into this hook (telling the hook the exact type of object to add and how it should be initialized).
+A custom hook that adds on `useList` the ability to dynamically add, move and remove list tems.
```
Comment (attila-berki) [inline packages/react-form/README.md:641]:
```suggestion
A custom hook that adds on `useList` the ability to dynamically add, move and remove list items.
```

What is the meaning of `moving a list` item?

File: packages/react-form/README.md [hunk 8]
```diff
@@ -698,88 +731,65 @@ You can choose to initialize the list with an existing number of cards or no car
...
 
 ```tsx
-const {submit, dirty, submitting, reset} = useForm({
+const form = useForm({
   fields: {
-    title: useField('')
-  },
-  dynamicLists: {
-    customerCards: useDynamicList<Card>([{cardNumber: '4242 4242 4242 4242', cvv: '422'}], emptyCardFactory)
+    customerCards: fields, // the fields destructured from useDynamicList
```
Comment (attila-berki) [inline packages/react-form/README.md:763]: I would add the `useDynamicList` detructuring above, it would be easier to understand what we mean here.

File: packages/react-form/README.md [hunk 9]
```diff
@@ -698,88 +731,65 @@ You can choose to initialize the list with an existing number of cards or no car
...
   onSubmit: async fieldValues => {
     console.log(fieldValues);
-    return {status: 'success'};
+    return submitSuccess();
   },
 });
 ```
 
-you can directly pass a dynamic list within `fields` but note that the `reset` function returned by `useForm` will only reset each field, plus the `dirty` state will only reflect dirty fields within each item.
+Or you can pass your dynamic lists within `dynamicLists`. This way the form `reset` and `dirty` will work as expected. Note that you still need to have one "classic field", `fields` cannot be optionnal at the moment. If you use TypeScript, you might want to pass generic types to `useForm` in order to make sure TypeScript will infer the `fieldValues` types correctly within `onSubmit`.
```
Comment (attila-berki) [inline packages/react-form/README.md:772]:
```suggestion
Or you can pass your dynamic lists within `dynamicLists`. This way the form `reset` and `dirty` will work as expected. Note that you still need to have one "classic field", `fields` cannot be optional at the moment. If you use TypeScript, you might want to pass generic types to `useForm` in order to make sure TypeScript will infer the `fieldValues` types correctly within `onSubmit`.
```

File: packages/react-form/README.md [hunk 10]
```diff
@@ -795,6 +805,7 @@ const form = useForm(config);
 - `config`, An object has the following fields:
 
   - `fields`, A dictionary of `Field` objects, dictionaries of `Field` objects, and lists of dictionaries of `Field` objects. Generally, you'll want these to be generated by the other hooks in this package, either `useField` or `useList`. This will be returned back out as the `fields` property of the return value.
+  - `useDynamicList`, optionnal dictionaries of `DynamicList`.
```
Comment (attila-berki) [inline packages/react-form/README.md:808]:
```suggestion
  - `useDynamicList`, optional dictionaries of `DynamicList`.
```

File: packages/react-form/src/hooks/form.ts [hunk 11]
```diff
@@ -18,7 +18,7 @@ import {useSubmit} from './submit';
 import {useDynamicListDirty, useDynamicListReset} from './list';
 
 /**
- * A custom hook for managing the state of an entire form. `useForm` wraps up many of the other hooks in this package in one API, and when combined with `useField` and `useList`, allows you to easily build complex forms with smart defaults for common cases.
+ * A custom hook for managing the state of an entire form. `useForm` wraps up many of the other hooks in this package in one API, and when combined with `useField` and `useList` and `useDynamicList`, allows you to easily build complex forms with smart defaults for common cases.
```
Comment (attila-berki) [inline packages/react-form/src/hooks/form.ts:21]:
```suggestion
 * A custom hook for managing the state of an entire form. `useForm` wraps up many of the other hooks in this package in one API, and when combined with `useField`, `useList` or `useDynamicList`, allows you to easily build complex forms with smart defaults for common cases.
```

File: packages/react-form/src/hooks/form.ts [hunk 12]
```diff
@@ -55,22 +51,16 @@ import {useDynamicListDirty, useDynamicListReset} from './list';
...
- *       <button type="button" onClick={addItem}>Add Payment Method</button>
  *     </form>
  *  );
  *```
  *
  * @param fields - A dictionary of `Field` objects, dictionaries of `Field` objects, and lists of dictionaries of `Field` objects. Generally, you'll want these to be generated by the other hooks in this package, either `useField` or `useList`. This will be returned back out as the `fields` property of the return value.
  *
  * @param onSubmit - An async function to handle submission of the form. If this function returns an object of `{status: 'fail', error: FormError[]}` then the submission is considered a failure. Otherwise, it should return an object with `{status: 'success'}` and the submission will be considered a success. `useForm` will also call all client-side validation methods for the fields passed to it. The `onSubmit` handler will not be called if client validations fails.
- * @param dynamicLists - Pass in dynamic lists objects into the form
+ * @param dynamicLists - optionnal dictionaries of `DynamicList`.
```
Comment (attila-berki) [inline packages/react-form/src/hooks/form.ts:63]:
```suggestion
 * @param dynamicLists - optional dictionaries of `DynamicList`.
```

**Review by attila-berki — APPROVED** [review]

File: packages/react-form/README.md [hunk 13]
```diff
@@ -698,88 +731,65 @@ You can choose to initialize the list with an existing number of cards or no car
         value={field.cvv.value}
```
[the remark this answers is not in the document]
Comment (sylvhama) [reply packages/react-form/README.md:731]: There is an example about that for `useField`, I think it's good to show what a field is made of.

File: packages/react-form/README.md [hunk 14]
```diff
@@ -638,54 +638,87 @@ function MyComponent() {
...
-This offers the same functionality as useList. `useDynamicList` is a hook that adds on `useList` by adding the ability to dynamically add and remove list items. The same way you would utilize `useList` is the same way you would utilize `useDynamicList` except some differences such as the `addItem`, and the `removeItem` function. We also have to pass in a factory into this hook (telling the hook the exact type of object to add and how it should be initialized).
+A custom hook that adds on `useList` the ability to dynamically add, move and remove list items.
 
 ##### Using `useDynamicList`
 
-Lets simulate having a user interface that allows you add as many payments methods as you wish.
+Let's simulate a user interface that allows you add as many credit cards as you wish.
 
-1. We have to initialize `useDynamicList` the following way
+You can initialize `useDynamicList` the following way
```
Comment (atesgoral) [inline packages/react-form/README.md:647]:
```suggestion
You can initialize `useDynamicList` the following way:
```

File: packages/react-form/README.md [hunk 15]
```diff
@@ -638,54 +638,87 @@ function MyComponent() {
...
 
 ```tsx
+interface Card {
+  id: string;
+  cardNumber: string;
+  cvv: string;
+}
+
 const emptyCardFactory = (): Card => ({
+  id: new Date.now().toString(),
```
Comment (atesgoral) [inline packages/react-form/README.md:657]:
```suggestion
  id: Date.now().toString(),
```

File: packages/react-form/README.md [hunk 16]
```diff
@@ -638,54 +638,87 @@ function MyComponent() {
...
 
 const {fields, addItem, removeItem, moveItem, reset, dirty} = useDynamicList(
-  [{cardNumber: '4242 4242 4242 4242', cvv: '000'}],
+  [],
   emptyCardFactory,
 );
 ```
 
-We can also have a factory that produces multiple cards such as
+You can also have a factory that produces multiple cards such as
```
Comment (atesgoral) [inline packages/react-form/README.md:668]:
```suggestion
You can also have a factory that produces multiple cards such as:
```

File: packages/react-form/README.md [hunk 17]
```diff
@@ -638,54 +638,87 @@ function MyComponent() {
...
-  {
-    cardNumber: '',
-    cvv: '',
-  },
-  {
-    cardNumber: '',
-    cvv: '',
-  },
+function generateId() {
+  return new Date.now().toString();
```
Comment (atesgoral) [inline packages/react-form/README.md:672]:
```suggestion
  return Date.now().toString();
```

File: packages/react-form/README.md [hunk 18]
```diff
@@ -638,54 +638,87 @@ function MyComponent() {
...
+  ];
+};
+
+const {fields, addItem, removeItem, moveItem, reset, dirty} = useDynamicList(
+  [],
+  emptyCardFactory,
+);
+```
+
+AddItem can accept an argument which will be passed to the factory. In this case, the factory could look like
```
Comment (atesgoral) [inline packages/react-form/README.md:688]:
```suggestion
AddItem can accept an argument which will be passed to the factory. In this case, the factory could look like:
```

File: packages/react-form/README.md [hunk 19]
```diff
@@ -698,88 +731,67 @@ You can choose to initialize the list with an existing number of cards or no car
...
 ```
 
-We render our UI representation of the fields by utilizing the `fields.map` function. For each field, we can use the handlers such as onChange, value, onBlur. These are the same handlers that useList provides.
+##### How to use it with `useForm`
 
-We can also utilize the `removeItem`, and `addItem` functions. In this example, these functions are attached to a button press. When we remove an item we pass in the index (to indicate what field item to remove). When we add a item, we utilize the factory we passed in when initializing `useDynamicList`.
+You can use `useDynamicList` with `useForm` in two ways:
 
-Note that addItem can also take in an argument which can then be passed onto the factory. In this case, the factory can be
+You can directly pass one or several dynamic lists within `fields`. But note that the `reset` function returned by `useForm` won't reset the dynamic list to its initial value. It will only reset the value of each field for each item. So for example if you remove an item, clicking on reset won't re-add it. Plus the adding or removing a field won't make the form dirty. The form will only be dirty if you edit the value of one item's field.
```
Comment (atesgoral) [inline packages/react-form/README.md:758]:
```suggestion
You can directly pass one or several dynamic lists within `fields`. But note that the `reset` function returned by `useForm` won't reset the dynamic list to its initial value. It will only reset the value of each field for each item. So, if you remove an item, clicking on reset won't re-add it. Plus, adding or removing a field won't make the form dirty. The form will only be dirty if you edit the value of an item's field.
```

File: packages/react-form/README.md [hunk 20]
```diff
@@ -698,88 +731,67 @@ You can choose to initialize the list with an existing number of cards or no car
...
   onSubmit: async fieldValues => {
     console.log(fieldValues);
-    return {status: 'success'};
+    return submitSuccess();
   },
 });
 ```
 
-you can directly pass a dynamic list within `fields` but note that the `reset` function returned by `useForm` will only reset each field, plus the `dirty` state will only reflect dirty fields within each item.
+Or you can pass your dynamic lists within `dynamicLists`. This way the form `reset` and `dirty` will work as expected. Note that you still need to have one "classic field", `fields` cannot be optional at the moment. If you use TypeScript, you might want to pass generic types to `useForm` in order to make sure TypeScript will infer the `fieldValues` types correctly within `onSubmit`.
```
Comment (atesgoral) [inline packages/react-form/README.md:774]:
```suggestion
Or you can pass your dynamic lists within `dynamicLists`. This way the form `reset` and `dirty` will work as expected. Note that you still need to have one "classic field", `fields` cannot be optional at the moment. If you use TypeScript, you might want to pass generic types to `useForm` to make sure TypeScript infers the `fieldValues` types correctly within `onSubmit`.
```

File: packages/react-form/README.md [hunk 21]
```diff
@@ -698,88 +731,67 @@ You can choose to initialize the list with an existing number of cards or no car
...
+};
+
+const customerCards = useDynamicList<Card>([], emptyCardFactory);
+
+const form = useForm<typeof fields, typeof customerCards>({
+  fields,
+  dynamicLists: {
+    customerCards,
   },
   onSubmit: async fieldValues => {
```
Comment (atesgoral) [inline packages/react-form/README.md:788]:
```suggestion
  onSubmit: async (fieldValues) => {
```

**Review by atesgoral — APPROVED** [review]
